### DISCUSSION TITLE
UP / Down event on Rekey

### DISCUSSION THREAD

**Discussion by YafimK** [strongswan/strongswan#1273] [Q&A]
Hey,
we've noticed that we are receiving duplicate UP and down events once we subscribe to **ike-updown** and **child-updown
**.
I've tracked down an issue marked as resolved in the old bug tracking system https://wiki.strongswan.org/issues/853 and I wonder if this was fixed also for the vici protocol events or not. Or is it a new bug? 

Thanks ahead

**Comment by tobiasbrunner** [comment]
What do you mean with "duplicate"?

**Reply by YafimK** [reply]
since we use "make before break" we got up event on the "new" tunnel and "down" event (i.e. with "IKE DELETE" in "active task queue") on the old one ... Same we had with child tunnel but in this case of course, the number of the child tunnel was increamented).
I will try to fetch the raw event and add it.

**Reply by tobiasbrunner** [reply]
That's of course absolutely normal. Reauthentcation is NOT rekeying, new SAs are created from scratch.

**Reply by YafimK** [reply]
So when we see the same duplications for CHILD SA, it means that its part of the reauthnication of IKEV2 tunnel and therefore expected? 
Is there a way within the protocol to identify that this is the case? or we have to keep state, for example tracking the UNIQUE ID field and in case a lower ID is DOWN then it should be ignored? 

an example for such raw events : 
`    "sa": {
      "dh-group": "ECP_256",
      "encr-alg": "AES_CBC",
      "encr-keysize": "128",
      "initiator-spi": "c08542099766169d",
      "integ-alg": "HMAC_SHA2_384_192",
      "local-host": "169.254.2.121",
      "local-id": "169.254.2.121",
      "local-port": "4500",
      "nat-any": "yes",
      "nat-local": "yes",
      "nat-remote": "yes",
      "prf-alg": "PRF_HMAC_SHA2_384",
      "remote-eap-id": "fakJBjyAAGMUYagO",
      "remote-host": "52.18.53.99",
      "remote-id": "172.31.3.204",
      "remote-port": "4500",
      "remote-vips": [
        "10.255.2.5"
      ],
      "responder-spi": "256e6f4e3eea80bc",
      "state": "DELETING",
      "tasks-active": [
        "IKE_DELETE"
      ],
      "uniqueid": "9",
      "version": "2"
    },
    "time": "2022-09-05T11:46:04Z"
  },`

and the "new" one - 
`"sa": {
      "dh-group": "ECP_256",
      "encr-alg": "AES_CBC",
      "encr-keysize": "128",
      "established": "0",
      "initiator-spi": "3d29534fb903aee4",
      "integ-alg": "HMAC_SHA2_384_192",
      "local-host": "169.254.2.121",
      "local-id": "169.254.2.121",
      "local-port": "4500",
      "nat-any": "yes",
      "nat-local": "yes",
      "nat-remote": "yes",
      "prf-alg": "PRF_HMAC_SHA2_384",
      "remote-eap-id": "fakJBjyAAGMUYagO",
      "remote-host": "52.18.53.99",
      "remote-id": "172.31.3.204",
      "remote-port": "4500",
      "responder-spi": "b68b745ced08d5e6",
      "state": "ESTABLISHED",
      "tasks-passive": [
        "IKE_AUTH",
        "IKE_CERT_POST",
        "IKE_CONFIG",
        "CHILD_CREATE",
        "IKE_AUTH_LIFETIME",
        "IKE_MOBIKE"
      ],
      "uniqueid": "10",
      "version": "2"
    },
    "time": "2022-09-05T11:46:41Z"
  },`

**Reply by tobiasbrunner** [reply]
> So when we see the same duplications for CHILD SA, it means that its part of the reauthnication of IKEV2 tunnel and therefore expected?

Maybe, but duplicates could theoretically also get created for other reasons (as initiator newer versions block that, but peers are free to create as many duplicate CHILD_SAs as they want).

> Is there a way within the protocol to identify that this is the case?

There isn't, especially as responder. The initiator just creates a new IKE_SA and CHILD_SAs, there is nothing on the protocol level that ties them together. If you want a more predictable behavior, use rekeying instead of reauthentication.

> or we have to keep state, for example tracking the UNIQUE ID field and in case a lower ID is DOWN then it should be ignored?

Possibly.

**Reply by YafimK** [reply]
Just to clarify - we are using the latest version 5.9.7 and we are the initator, since in must cases the other side (either AWS, meraki and etc. are usually set as responders.
But I did reread https://docs.strongswan.org/docs/5.9/config/rekeying.html and I got the point that we don't really have to set REAUTH with IkeV2 based tunnels ... and we can use just rely on rekey as you suggested. 

Thanks!
